### DISCUSSION TITLE
How to handle multiple sets of env variables in one app?

### DISCUSSION THREAD

**Discussion by noemi-salaun** [symfony/symfony#43113] [Q&A] [answered]
Hello, I would like to start a conversation about how we should properly configure multiple sets of environment variables in a single app. When I say a single app, I mean one directory deployed on a server.

My app offers a different version depending of the language. It support 10 languages, each with its own domain, its own database and its own vhost on the server.

All these versions used the same compiled kernel container so I need to use env variables to configure the database, the facebook app id, the timezone, the mailer "from" name, etc.

My first idea was to put all these env variables in the nginx vhost configuration. It worked well for the web part of the application, but I also need the language configuration to be available for crons or utility scripts.

My actual solution is to have a php file for each language for the staging and the prod environment. These files have the same format as the `env.local.php` file and can be included during the bootstrap. They are stored in a directory inside my project

- my-project-dir/
    - env/
        - staging/
            -  fr.php
            -  es.php
            -  it.php
            -  ru.php
            -  ...
        - prod/
            -  fr.php
            -  es.php
            -  it.php
            -  ru.php
            -  ...

And they are included like that in the bootstrap

```php
$env = $_SERVER['APP_ENV'];
$lang = $_SERVER['APP_LANGUAGE_ID'];

if (is_array($array = @include dirname(__DIR__)."/env/$env/$lang.php")) {
    $dotEnv->populate($array);
}
```

Now the `env.local.php` only need to have `APP_ENV=prod` and I can call the Symfony console like that

```bash
APP_LANGUAGE_ID=fr php bin/console my:script
```

It works well, but with this kind of configuration, I always need a custom boostrap file and I can't use the new Runtime component, or the secret vault. And also I can't use simple container parameter to configure things like Facebook App ID, because if it's not an env variable, it will be the same value for every version once the kernel container is built.

Do I missed some Symfony capabilities, or some standard Linux tools for handling environment variables ?
I need your help 😃

**Comment by upyx** [comment]
At first, I think it would be a good idea to separate containers by the environment. To do it, you can extend a kernel, add a language ID, and override the `getCacheDir()` method.

```php
class Kernel extends BaseKernel
{
    use MicroKernelTrait;

    private $langId;

    public function __construct(string $environment, bool $debug, string $langId)
    {
        parent::__construct($environment, $debug);

        $this->langId = $langId;
    }

    public function getCacheDir(): string
    {
        return parent::getCacheDir() . '/' . $this->langId;
    }

    /* ... */
}
```

That will allow you to use different configuration files for various languages with any structure of directories and files. For example:

```php
    /* ... */

    protected function configureContainer(ContainerConfigurator $container): void
    {
        /* other imports ... */

        $container->import('../config/{parameters}/' . $this->environment . '_' . $this->langId . '.php');

        /* other imports ... */
    }

    /* ... */
```

Optionally, you can override the `getLogDir()`  method to separate logs loo.

```php
    /* ... */

    public function getLogDir(): string
    {
        return parent::getLogDir() . '/' . $this->langId;
    }

    /* ... */
```


Then, the Runtime component has been developed right for managing a bootstrapping process.

You can set the `dotenv_path` option to select what the "env" file use.

```php
use App\Kernel;

$_SERVER['APP_RUNTIME_OPTIONS'] = [
    'dotenv_path' => 'env/' . $_SERVER['APP_ENV'] . '/' . $_SERVER['APP_LANGUAGE_ID'] . '.env',
];

require_once dirname(__DIR__) . '/vendor/autoload_runtime.php';

return static function (array $context) {
    return new Kernel($context['APP_ENV'], (bool)$context['APP_DEBUG'], $context['APP_LANGUAGE_ID']);
};
```

Or you can disable the "env" loading, and propogate it on your own.
```php
use App\Kernel;
use Symfony\Component\Dotenv\Dotenv;

$_SERVER['APP_RUNTIME_OPTIONS'] = [
    'disable_dotenv' => true,
];

require_once dirname(__DIR__) . '/vendor/autoload_runtime.php';

return static function (array $context) {
    (new Dotenv())->bootEnv('env/' . $context['APP_ENV'] . '/' . $context['APP_LANGUAGE_ID'] . '.env');

    return new Kernel($context['APP_ENV'], (bool)$context['APP_DEBUG'], $context['APP_LANGUAGE_ID']);
};
```

Check the https://symfony.com/doc/current/components/runtime.html#using-options for other ideas :wink:

**Reply by noemi-salaun** [reply]
That's what I tried the first time, but then if I want to separate my containers I will need to warmup the app 10 times for each deployment and I won't be able to use php 7.4 preloading because it can only be one app in the preload cache.

Additionnaly, tools like Symfony Insight does not like custom kernel with a modified constructor signature :/

**Reply by upyx** [reply]
Yep. It's familiar for me :) And you **already** have 10 applications on different vhosts with different databases. Think about separate their code and PHP-FPM pools too. Having the possibility to deploy them separately isn't a bad thing.

Nevertheless, changing the signature of the kernel is not required. You can use the env variable inside the constructor or use a setter outside. If you need different sets of parameters, you would have to warm up them all. Otherwise, you can rely on env parameters only. In that case, you might use the Runtime component to simplify bootstrapping.

**Reply by faizanakram99** [reply]
I don't see a reason for multiple cache folders, we host a multi tenant app (single web server, multiple database) where each tenant used their own set of env vars with a single cached container.

This is the beauty of env vars over kernel parameters, it is evaluated at runtime.

**Reply by faizanakram99** [reply]
We require env vars based on http_host by using a custom env loader.

See EnvVarLoaderInterface https://github.com/symfony/symfony/pull/34295

**Reply by noemi-salaun** [reply]
@faizanakram99 I will look at the `EnvVarLoaderInterface`. How do you handle the loading of env vars for bin/console ?

**Reply by faizanakram99** [reply]
Basically, we fetch tenant id from http_host for web processes and then rest of application depends upon tenant id.
For symfony commands, tenant id is set before running any command.

TENANT_ID=some-tenant ./bin/console some-command.

I'm away from desk right now, will try to post our whole setup tomorrow.

**Reply by noemi-salaun** [reply]
Ok, I do the same thing with my `APP_LANGUAGE_ID`.

> I'm away from desk right now, will try to post our whole setup tomorrow.

It would be awesome :)

**Reply by faizanakram99** [reply]
added an answer below

**Comment by stof** [comment]
> And also I can't use simple container parameter to configure things like Facebook App ID, because if it's not an env variable, it will be the same value for every version once the kernel container is built.

That's totally expected. Any parameter that is not accessed through `%env(...)%` is not dynamic and requires clearing the cache to change it

**Reply by noemi-salaun** [reply]
Yeah it sounds obvious once I take a step back and think about it, but at that time I wasn't aware of all the details of the compiled kernel container

**Comment by faizanakram99** [comment, marked as answer]
The idea is to fetch env vars based on APP_LANGUAGE_ID.

We do something similar and here is our setup.

We use a custom runtime which extends SymfonyRuntime and we set an env var INSTALLATION_NAME which is similar to APP_LANGUAGE_ID there if it is not already set (based on the hostname of subdomain).

Afterwards we load env vars using a custom env var loader (any service implementing EnvVarLoaderInterface) to load env vars of that particular tenant (DATABASE_URL, etc).

These tenant specific env vars are basically stored in AWS secrets manager (earlier we used sops to manage them, see mozilla/sops github repository) and in our CI server we dump these env vars into separate PHP files named like `.env.{INSTALLATION_NAME}.php` for performance reasons (as PHP files are opcached and are as fast as accessing something from memory)

For cli commands we supply INSTALLATION_NAME just before running command to make sure command runs for that particular tenant.

Profit.

Here is the gist with some of the files, that should give u some hint.

https://gist.github.com/faizanakram99/e61639b4fea0093af81a67b8d6325598

**Reply by noemi-salaun** [reply]
Thanks, amazing!

I don't see how your `TenantEnvLoader` is instanciated in your gist, there is only access to its constants

**Reply by faizanakram99** [reply]
Yep, it is a service (custom env var loader) instantiated by symfony automatically. Whatever is returned by `EnvVarLoaderInterface::loadEnvVars()` will be automatically registered as env var in Symfony DIC.

**Reply by faizanakram99** [reply]
**NOTE:** Those env vars won't be available in `$_SERVER` or `$_ENV`, instead they can be used in code by referencing them with `%env('SOMETHING')%` in config files and injecting them wherever needed.

**Reply by noemi-salaun** [reply]
Oww ok, it sounds great! Thank you very much for your time and explanation :)

**Reply by faizanakram99** [reply]
Glad it helped
